File: spark_scala_study/src/main/scala/com/zhangzl/study/core/AccumulatorVariable.scala
package com.zhangzl.study.core

import org.apache.spark.{SparkConf, SparkContext}

/**
  * Description  : 共享变量之累加变量
  * Class   Path : com.zhangzl.study.core
  * Create  User : zhangzuolong
  * Create  Time : 2017/2/14 14:28
  * Project Name : technology_sophisticated_study
  */
object AccumulatorVariable {
    def main(args: Array[String]): Unit = {

        val conf = new SparkConf().setAppName("AccumulatorVariable").setMaster("local")
        val sc = new SparkContext(conf)

        /**
          * 定义共享变量之累加变量
          */
        val accumulator = sc.accumulator(0)
        val initRdds = sc.parallelize(Array(1,2,3,4,5,6,7),1)
        initRdds.foreach(number => accumulator.add(number))
        println("total == "+ accumulator)
    }

}

File: spark_java_study/src/main/java/com/zhangzl/study/core/AccumulatorVariable.java
package com.zhangzl.study.core;

import org.apache.spark.Accumulator;
import org.apache.spark.SparkConf;
import org.apache.spark.api.java.JavaRDD;
import org.apache.spark.api.java.JavaSparkContext;
import org.apache.spark.api.java.function.VoidFunction;

import java.util.Arrays;
import java.util.List;

/**
 * Description  : 共享变量之累加变量
 * Class   Path : com.zhangzl.study.core
 * Create  User : zhangzuolong
 * Create  Time : 2017/2/13 18:07
 * Project Name : technology_sophisticated_study
 */
public class AccumulatorVariable {

    public static void main(String[] args) {


        SparkConf conf = new SparkConf().setAppName("AccumulatorVariable").setMaster("local");

        JavaSparkContext sc = new JavaSparkContext(conf);

        final Accumulator<Integer>  total = sc.accumulator(0);
        List<Integer> numberList = Arrays.asList(1,2,3,4,5,6,7,8,9);

        JavaRDD<Integer> numbers = sc.parallelize(numberList);
        numbers.foreach(new VoidFunction<Integer>() {
            public void call(Integer v1) throws Exception {
                total.add(v1);
            }
        });

        System.out.println("total == "+total);
        sc.close();

    }
}
